scripts/irods/test/test_icp.py

```diff
@@ -51,3 +50,26 @@ def test_icp_closes_file_descriptors__4862(self):
         finally:
             shutil.rmtree(test_dir_path, ignore_errors=True)
 
+    def test_multithreaded_icp__issue_5478(self):
+        def do_test(size, threads):
+            file_name = 'test_multithreaded_icp__issue_5478'
+            local_file = os.path.join(self.alice.local_session_dir, file_name)
+            dest_path = os.path.join(self.alice.session_collection, file_name)
+            another_dest_path = os.path.join(self.alice.session_collection, file_name + '_copy')
+
+            try:
+                lib.make_file(local_file, size)
+
+                self.alice.assert_icommand(['iput', local_file, dest_path])
+
+                cp_thread_count = self.alice.run_icommand(['icp', '-v', dest_path, another_dest_path])[0].split('|')[2].split()[0]
+
+                self.assertGreaterEqual(int(cp_thread_count), threads)
+
+            finally:
+                self.alice.assert_icommand(['irm', '-f', dest_path])
+                self.alice.assert_icommand(['irm', '-f', another_dest_path])
+
+        do_test(0, 0)
+        do_test(31457280, 1) # 30MiB
+        do_test(52428800, 1) # 50MiB
```

Comment:
Any suggestions on what numbers to check for here? I remember that there is a place where we calculate this kind of thing but I do not now recall where that is. Also, 0 is a little special in that I might need a `self.assertEqual` rather than `self.assertGreaterEqual`, but if we can determine the exact number of threads expected in each case then I can always use `self.assertEqual`.

Reply:
and `msiSetNumThreads` has the real logic

Reply:
Yep, here's the magic formula...

https://github.com/irods/irods/blob/9eb6c23df45cdedff8ee9c3af71a65d304635037/server/re/src/reSysDataObjOpr.cpp#L828